File: collectors/services.py
import json
from datetime import datetime, timezone

def collect(hostname):
    ts = datetime.now(timezone.utc)
    rows = []
    try:
        import wmi
        c = wmi.WMI()
        for svc in c.Win32_Service():
            rows.append((
                ts, hostname,
                svc.Name,
                svc.DisplayName,
                svc.State,
                svc.StartMode,
                int(svc.ProcessId) if svc.ProcessId else None,
                json.dumps({"Name": svc.Name, "State": svc.State, "StartMode": svc.StartMode, "Status": getattr(svc, 'Status', None)})
            ))
    except Exception as e:
        # fallback via psutil
        try:
            import psutil
            for svc in psutil.win_service_iter():
                try:
                    info = svc.as_dict()
                    rows.append((ts, hostname, info.get('name'), info.get('display_name'), info.get('status'), info.get('start_type'), int(info.get('pid')) if info.get('pid') else None, json.dumps(info, default=str)))
                except:
                    continue
        except Exception as e2:
            rows.append((ts, hostname, "error", str(e2)[:200], "error", None, None, json.dumps({"error": str(e)})))
    columns = ["ts","hostname","name","display_name","status","start_type","pid","raw"]
    return columns, rows

File: go/internal/collectors/services.go
package collectors

import (
	"encoding/json"
	"time"

	"github.com/yusufpapurcu/wmi"
)

// win32Service mirrors WMI Win32_Service minimal fields.
type win32Service struct {
	Name        string
	DisplayName string
	State       string
	StartMode   string
	ProcessId   uint32
	Status      string
}

// CollectServices mirrors collectors.psutil_collectors.collect_services via WMI.
func CollectServices(hostname string, ts time.Time) (Result, error) {
	cols := []string{"ts", "hostname", "name", "display_name", "status", "start_type", "pid", "raw"}
	var dst []win32Service
	err := wmi.Query("SELECT Name, DisplayName, State, StartMode, ProcessId, Status FROM Win32_Service", &dst)
	rows := [][]any{}
	if err == nil {
		for _, svc := range dst {
			var pid any
			if svc.ProcessId != 0 {
				pid = int64(svc.ProcessId)
			}
			raw, _ := json.Marshal(map[string]any{"Name": svc.Name, "State": svc.State, "StartMode": svc.StartMode, "Status": svc.Status})
			rows = append(rows, []any{
				ts, hostname, svc.Name, svc.DisplayName, svc.State, svc.StartMode, pid, json.RawMessage(raw),
			})
		}
	}
	return Result{Table: "monitor.services", Columns: cols, Rows: rows}, nil
}
